src/qpx_writer.cpp

```diff
@@ -0,0 +1,415 @@
+  dev->rd_buf[foffs + 3] = 0x14; // for lead-in data form = 0x14
+  // lead-out end addr
+  dev->rd_buf[foffs + 5] = lout / 4500;
+  dev->rd_buf[foffs + 6] = (lout / 75) % 60;
+  dev->rd_buf[foffs + 7] = lout % 75;
+  /*
+      printf("Lead-Out pos: %02d:%02d.%02d\n",
+                      dev->rd_buf[foffs+5],
+                      dev->rd_buf[foffs+6],
+                      dev->rd_buf[foffs+7]
+                      );
+  */
+  foffs += 8;
+
+  dev->cmd[0] = MMC_SEND_CUE_SHEET;
+  dev->cmd[6] = (foffs >> 16) & 0xFF;
+  dev->cmd[7] = (foffs >> 8) & 0xFF;
+  dev->cmd[8] = foffs & 0xFF;
+
+  if ((dev->err = dev->cmd.transport(WRITE, dev->rd_buf, foffs))) {
+    sperror("SEND_CUE_SHEET", dev->err);
+    return (dev->err);
+  }
+  return 0;
+}
+
+int qpxwriter_cd::write_lead_in() {
+  memset(dev->rd_buf, 0, bufsz_rd);
+  int32_t lba = -150;
+  printf("Writing Lead-In...\n");
+  for (int i = 0; i < 6; i++) {
+    if (write_data(lba, 25))
+      return dev->err;
+    lba += 25;
+  }
+  return 0;
+}
+
+/*
+int qpxwriter_cd::open_session()
+{
+        return (set_write_parameters_def(0,0));
+}
+*/
+
+int qpxwriter_cd::open_track(uint32_t size) {
+#if (WTYPE_CD_DEFAULT == WTYPE_SAO)
+  printf("Writing in SAO mode\n");
+#elif (WTYPE_CD_DEFAULT == WTYPE_TAO)
+  printf("Writing in TAO mode\n");
+#endif
+  if (set_write_parameters_def(((dev->capabilities & CAP_BURN_FREE)), simul)) {
+    return 1;
+  }
+  /*
+      mode_sense(dev, 0x05, 0, 60);
+      for (int i=0; i<52; i++) {
+              if (!(i%8)) printf("\n");
+              printf(" %02X", dev->rd_buf[i+8]);
+      }
+  */
+#if (WTYPE_CD_DEFAULT == WTYPE_SAO)
+  if (send_cue_sheet(size))
+    return 1;
+  if (write_lead_in())
+    return 1;
+#endif
+  return 0;
+}
+
+int qpxwriter_cd::close_track() {
+  printf("Closing track...\n");
+  if (flush_cache(dev, true))
+    return 1;
+  wait_unit_ready(dev, 300, 1);
+
+#if (WTYPE_CD_DEFAULT == WTYPE_TAO)
+  close_track_session(dev, 1, CLOSE_TRACK); // close track
+#endif
+  /*
+      do {
+              usleep(20000);
+              get_buffer_capacity(dev,&ubuft,&ubuff);
+              printf("buf: %d %%\r", 100*(ubuft-ubuff) / ubuft);
+      } while (ubuff<ubuft);
+      printf("\n");
+  */
+  wait_unit_ready(dev, 120);
+  return 0;
+}
+
+int qpxwriter_cd::fixate() {
+  printf("Fixating...\n");
+#if (WTYPE_CD_DEFAULT == WTYPE_TAO)
+  close_track_session(dev, 1, CLOSE_SESSION); // close session
+#endif
+  if (flush_cache(dev, false))
+    return 1;
+  return (wait_fix(dev, 120));
+}
+
+//
+// DVD- functions...
+//
+int qpxwriter_dvdminus::open_session() {
+  bool bfree = ((dev->capabilities & CAP_BURN_FREE));
+  bfree = 1;
+  if (mode_sense(dev, MODE_PAGE_WRITE_PARAMETERS, 0, 60))
+    return -1;
+  if (!!(dev->rd_buf[8 + 2] & 0x40) == bfree) {
+    printf("BURN-Free is %s\n", (dev->rd_buf[8 + 2] & 0x40) ? "ON" : "OFF");
+  } else {
+    printf("Turning BURN-Free %s\n", bfree ? "ON" : "OFF");
```

Comment:
Codacy found an issue: [Condition 'bfree' is always true](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)